### ISSUE TITLE
Explore using Zookeeper instead of Eureka for JHipster Registry

### ISSUE THREAD

**Issue by deepu105** [jhipster/generator-jhipster#19401]
Explore using Zookeeper via the [Spring Cloud Zookeeper integration](https://cloud.spring.io/spring-cloud-zookeeper/reference/html/). On paper it seems like all the features that were offered by JHipster Registry can be provided by this. 

This ticket is for exploration and a possible POC

Some notable points from the docs

- With a few annotations, you can quickly enable and configure the common patterns inside your application and build large distributed systems with Zookeeper based components.
- The provided patterns include Service Discovery and Configuration. The project also provides client-side load-balancing via integration with Spring Cloud LoadBalancer.
- [Curator](https://curator.apache.org/)(A Java library for Zookeeper) provides Service Discovery through a [Service Discovery Extension](https://curator.apache.org/curator-x-discovery/). Spring Cloud Zookeeper uses this extension for service registration and discovery.
- Feign and Spring Cloud LoadBalancer also work with Spring Cloud Zookeeper.
- Spring Cloud Zookeeper implements the ServiceRegistry interface, letting developers register arbitrary services in a programmatic way.

**Comment by deepu105** [comment]
cc @gmarziou @PierreBesson @pascalgrimaud @mraible

**Comment by PierreBesson** [comment]
I don't think this would be hard to replace Eureka by Zookeeper as the discovery integration is handled by the Spring Cloud abstraction layer. It might actually be a good choice for long term support as Zookeeper is well maintained and supported by the Apache Foundation.
Although, if we go this route, it will really be going full circle. As the oldest service discovery systems I have known about were based on Zookeeper and Eureka was supposed to be the newer/better/easier way to do service discovery. 
Ping @juliensadaoui, you might have an opinion on this.

**Comment by mraible** [comment]
This comment seems relevant to this discussion:

> Eureka is not going away any time soon BTW

https://twitter.com/spencerbgibb/status/1557718706033147905

**Comment by atomfrede** [comment]
Zookeeper works well, we used it as configuration service. Solr uses it for discovery and configuration. Not sure why we should replace eureka. Both are quite mature with less frequent changes, but when it comes to central infrastructure I guess thats more of a pro argument :)

**Comment by deepu105** [comment]
For me JHipster registry would be good enough if Eureka was well
maintained. Currently that doesn't seem like the case. The last commit to
the repo was a year ago and there are even some security issues open. I
personally would love to have JHipster registry back in its full glory
(with the admin dashboard and stuff)

**Comment by atomfrede** [comment]
Would it be enough to use the spring eureka server application?

**Comment by gmarziou** [comment]
@atomfrede it's just a spring boot application with spring-boot-eureka-starter, similar to what the jhipster-registry is but without any UI. So if you idea is to use this starter as a replacement for jhipster-registry, yes it can work if you don't use other features, you just need a pom and one class to build it.

The benefit from using this starter is that Spring team updates vulnerable dependencies (e.g. xstream, jackson, ...) and so the CVEs that @deepu105 has seen in Netflix eureka core project are fixed in the starter, this can work forever as long as there is no breaking changes in dependencies or as long as there's no vulnerability in eureka core itself.

**Comment by atomfrede** [comment]
Just came across this baeldung post, it looks like there is some kind of ui https://www.baeldung.com/spring-cloud-netflix-eureka. For sure less powerful than the jhipster registry, but better than nothing.

**Comment by narendracode** [comment]
I integrated  Zookeeper using [curator](https://github.com/apache/curator) in one of my blockchain based projects. If that's something in option then I can help with it.

**Comment by cbornet** [comment]
I work in the same team as the Apache Curator PMC chair if that can help 😉

**Comment by eolivelli** [comment]
Apache ZooKeeper and Apache Curator are stable, well maintained and used by many small and huge companies in the world (Facebook, Twitter, Apple....).

I am not aware of the usecases here, if you need something for service discovery and configuration management I think that supporting ZK and Curator will be a good choice. I will be happy to help at least with reviews.

**Comment by deepu105** [comment]
I dont think I'll have the time to explore this anytime soon, so if someone
wants to do it, please feel free to explore

Thanks & Regards,
Deepu

**Comment by Sebas5207418** [comment]
Hi @deepu105, I am interested in exploring this and creating a POC for Zookeeper integration via Spring Cloud Zookeeper. I will start working on a proposal

**Comment by Sebas5207418** [comment]
Hello @deepu105 and JHipster team, I have successfully started the Zookeeper POC using Docker. Attached is the screenshot of the running container as proof of the local environment setup. Ready for the next steps regarding the $500 bug bounty

[c2655681-8d0f-4196-a669-1b61eea1a1ff.txt](https://github.com/user-attachments/files/25445789/c2655681-8d0f-4196-a669-1b61eea1a1ff.txt)

**Comment by mshima** [comment]
Unfortunately jhipster-registry is deprecated and not maintained anymore.
I think this should be closed @DanielFran.

**Comment by Abu1982** [comment]
Hi, I'd like to work on this issue.

I've reviewed the thread and confirmed there is no active PR for it.

**Planned approach:** Analyze and fix based on error signals in jhipster/generator-jhipster

I'll submit a focused PR once the fix is ready and locally verified.

**Comment by Abu1982** [comment]
Hi, I'd like to work on this issue.

I've reviewed the thread and confirmed there is no active PR for it.

**Planned approach:** Analyze and fix based on error signals in jhipster/generator-jhipster

I'll submit a focused PR once the fix is ready and locally verified.

**Comment by Abu1982** [comment]
Hi, I'd like to work on this issue.

I've reviewed the thread and confirmed there is no active PR for it.

**Planned approach:** Analyze and fix based on error signals in jhipster/generator-jhipster

I'll submit a focused PR once the fix is ready and locally verified.

**Comment by mshima** [comment]
We do have lots of old bug that are not relevant anymore.
JHipster Registry was deprecated and archived.
Closing.

**Comment by Sebas5207418** [comment]
Hai detto
Hi @deepu105 @PierreBesson @gmarziou,

I've submitted the full implementation for Zookeeper Service Discovery in PR #32599. It's production-ready, follows standard JHipster patterns (logic, dependencies, and cloud-ready YAML), and is ready for review.

Thanks!

**Comment by Sebas5207418** [comment]
@mshima @deepu105 @PierreBesson @gmarziou @DanielFran

This issue cannot be closed because the requested Zookeeper integration is now fully implemented in PR #32599—and it directly addresses the core problem you’ve highlighted:

### Why Zookeeper is Necessary (Not Optional)

* **JHipster Registry is Deprecated:** As @PierreBesson and @deepu105 noted in this thread, the JHipster Registry (Eureka-based) was deprecated with no maintained alternative. The official Eureka Docker image is years old, and the Netflix Eureka repo has open CVEs.
> "The problem with keeping Eureka support is that we have deprecated our own Eureka server app [...] and there are no alternative that I know of." — @PierreBesson

* **Zookeeper is the Requested Solution:** This PR implements exactly what was asked: a modern, maintained alternative to Eureka (Apache Zookeeper + Spring Cloud). It has no dependency on the deprecated Registry and is cloud-ready with environment variables support.

* **The Bounty was for a POC → This is Production-Ready:**
The issue asked for "exploration and a possible POC". PR #32599 delivers a full, tested implementation: Generator logic, Maven dependencies (Spring Cloud 4.x), Java config (@EnableDiscoveryClient), and YAML for dev/prod.

### What Closing This Issue Would Mean
Closing this now, after a complete implementation has been submitted, would ignore the urgent need for a maintained service discovery option and would undermine the trust in the JHipster Bug Bounty program ($500).

I respectfully request to **reopen this issue** and proceed with the review of PR #32599. It is the solution to the deprecation problem you've discussed here.

Thanks!

**Comment by mshima** [comment]
This issue was meant to investigate usage in JHipster Registry https://github.com/jhipster/jhipster-registry not generator-jhipster.
